tool/dart_model_generator/lib/generate_dart_model.dart

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:dart_style/dart_style.dart';
+import 'package:json_schema/json_schema.dart';
+
+class DartModelGenerator {
+  void run() {
+    final result = <String>[];
+    final schema = JsonSchema.create(
+        File('schemas/dart_model.schema.json').readAsStringSync());
+    for (final def in schema.defs.entries) {
+      result.add(_generateExtensionType(def.key, def.value));
+    }
+
+    File('pkgs/dart_model/lib/src/dart_model.g.dart').writeAsStringSync(
+        DartFormatter().formatSource(SourceCode(result.join('\n'))).text);
+  }
+
+  String _generateExtensionType(String name, JsonSchema definition) {
+    final result = StringBuffer();
+
+    final jsonType = switch (definition.type) {
+      SchemaType.object => 'Map<String, Object?> node',
+      SchemaType.string => 'String string',
+      _ => throw UnsupportedError('Schema type ${definition.type}.'),
+    };
+    result.writeln('extension type $name.fromJson($jsonType) {');
+
+    final propertyMetadatas = definition.properties.entries
+        .map((e) => _readPropertyMetadata(e.key, e.value))
+        .toList();
+    if (definition.type == SchemaType.object) {
```

Comment:
Should this be a switch so we can get exhaustiveness? I think it would also look nicer, and would encourage maybe breaking out some of the code into helper methods which would make it easier to follow (`_generateConstructor` etc which are more self documenting).